readers-frontend/src/components/NavbarColourSwitcher.astro

```diff
@@ -9,56 +9,173 @@ interface Props {
-    scrollCategories.forEach((category) => {
+<navbar-controller
+  data-categories={serialisedScrollCategories}
+  data-style={style}></navbar-controller>
+
+<script>
+  /**
+   * Custom element that controls Navbar colour changes
+   */
+  class NavbarController extends HTMLElement {
+    // Properties
+    scrollCategories: Map<string, string>;
+    assignedStyle: "muse" | "nouse" = "nouse";
+    desktopNav: HTMLElement | undefined;
+    mobileHeader: HTMLElement | undefined;
+    mobileNouseLogo: HTMLElement | undefined;
+    sections: NodeListOf<Element> | undefined;
+    observer: IntersectionObserver | undefined;
+
+    constructor() {
+      super();
+      // Get parameters
+      this.scrollCategories = new Map<string, string>(
+        JSON.parse(this.dataset.categories || "[]")
+      );
+
+      if (this.dataset.style === "muse" || this.dataset.style === "nouse")
+        this.assignedStyle = this.dataset.style;
+
+      // Get document elements
+      this.desktopNav = document.querySelector("#desktopNav") as HTMLElement;
+      this.mobileHeader = document.querySelector(
+        "#mobileHeader"
+      ) as HTMLElement;
+      this.mobileNouseLogo = document.querySelector(
+        "#mobileHeader img"
+      ) as HTMLElement;
+      this.sections = document.querySelectorAll("[id$=section]");
+
+      // Intersection Observer
+      this.observe();
+
+      // Window resize event
+      this.registerResizeListener();
+    }
+
+    /**
+     * Register a resize event listener on the window
+     */
+    registerResizeListener() {
+      // Resize event is needed to switch behaviour
+      // when user shrinks window, or when tablet is rotated
+      window.addEventListener("resize", (_) => {
+        this.observe();
+      });
+    }
+
+    /**
+     * Initialises observing elements to animate navbar colour change
+     */
+    observe() {
+      // Is this a mobile device?
+      const mobile = window.innerWidth <= 768;
+
+      // Get 1 rem size
+      const rem = parseInt(getComputedStyle(document.documentElement).fontSize);
+
+      // Desktop navbar is 3rem
+      const desktopMargin = window.innerHeight - 3 * rem;
+
+      // Mobile navbar header is 7 rem
+      const mobileMargin = window.innerHeight - 7 * rem;
+
+      // Disconnect previous observer if any
+      if (this.observer) this.observer.disconnect();
+
+      // Trigger intersection only when item is in view port
+      // and navbar height away from top of the page.
+      // Whole viewport is default intersection element.
+      const observerOptions = {
+        // top: 0, left: 0, bottom: window - navbarHeight, right: 0
+        rootMargin: `0px 0px -${mobile ? mobileMargin : desktopMargin}px 0px`,
+      };
+
+      // Create Intersection Observer
+      this.observer = new IntersectionObserver(
+        this.changeNavbarColour,
+        observerOptions
+      );
+
+      // Observe each section
+      this.sections!.forEach((section) => {
```

Comment:
We're using the `!` to have typescipt ignore the null check. I would say instead let's have a `if (this.sections)` statement instead